Chap_API_Server.tex

```diff
@@ -238,6 +243,251 @@ \subsection{\code{PMIx_server_register_nspace}}
+
+Upon conclusion of this step, the \refarg{info} array might look like:
+
+\begingroup
+\begin{figure*}[ht!]
+  \begin{center}
+    \includegraphics[clip,width=0.4\textwidth]{figs/jobinfo.pdf}
+  \end{center}
+  \caption{Job-level information elements}
+  \label{fig:jobinfo}
+\end{figure*}
+\endgroup
+
+Note that in this example, \refattr{PMIX_NUM_NODES} is not required as that information is contained in the \refattr{PMIX_NODE_MAP} attribute. Similarly, \refattr{PMIX_JOB_SIZE} is not technically required as that information is contained in the \refattr{PMIX_PROC_MAP} when combined with the corresponding node map - however, there is no issue with including the job size as a separate entry.
+
+\item Application-level information includes all application-specific values such as \refattr{PMIX_APP_SIZE} and \refattr{PMIX_APPLDR}. If the \refterm{job} contains only a single \refterm{application}, then the application-specific values can be specified independently - i.e., in their own \refstruct{pmix_info_t} elements of the \refarg{info} array - or as a \refstruct{pmix_data_array_t} array of \refstruct{pmix_info_t} using the \refattr{PMIX_APP_INFO_ARRAY} attribute and identifed by including the \refattr{PMIX_APPNUM} attribute in the array. Use of the array format is  \emph{required} in cases where non-specific attributes (e.g., \refattr{PMIX_NODE_MAP}) are passed to describe aspects of the application.
+
+However, in the case of a job consisting of multiple applications, all application-specific values for each application \emph{must} be provided using the \refattr{PMIX_APP_INFO_ARRAY} format, each identified by its \refattr{PMIX_APPNUM} value.
+
+Upon conclusion of this step, the \refarg{info} array might look like that shown in \ref{fig:appinfo}, assuming there are two applications in the job being registered:
+
+\begingroup
+\begin{figure*}[ht!]
+  \begin{center}
+    \includegraphics[clip,width=0.5\textwidth]{figs/appinfo.pdf}
+  \end{center}
+  \caption{Application-level information elements}
+  \label{fig:appinfo}
+\end{figure*}
+\endgroup
+
+\item Process-level information includes an entry for each process in the job being registered, each entry marked with the \refattr{PMIX_PROC_DATA} attribute. The \refterm{rank} of the process \emph{must} be the first entry in the array - this provides efficiency when storing the data. Upon conclusion of this step, the \refarg{info} array might look like the diagram in \ref{fig:procinfo}:
+
+\begingroup
+\begin{figure*}[ht!]
+  \begin{center}
+    \includegraphics[clip,width=0.5\textwidth]{figs/procinfo.pdf}
+  \end{center}
+  \caption{Process-level information elements}
```

Comment:
On the picture there is a "Node-Rank" and "Node-ID"
What is the difference?

Reply:
The labels are attributes - e.g.., "Node-Rank" => PMIX_NODE_RANK. The node rank is the rank of the process looking at all procs on the node, regardless of job. Node iD identifies the node the process is executing upon.